File: packages/backend/src/bot/core/map.ts
import type * as Table from "@prisma/client";
import { assertIsDefined } from "./assert-defined";
import { Emote } from "./emotes";
import { TrackmaniaTime } from "./time";

export enum Medal {
	CHAMPION = 0,
	AUTHOR = 1,
	GOLD = 2,
	SILVER = 3,
	BRONZE = 4,
	NONE = 5,
}

const stylingChars = ["w", "n", "o", "i", "t", "s", "g", "z", "$", "W", "N", "O", "I", "T", "S", "G", "Z"];
const stylingRegex = new RegExp(`[$]([${stylingChars.join("")}]|[0-9A-Fa-f]{3})`, "g");

export class TrackmaniaMap {
	constructor(public readonly data: Table.TrackmaniaMap) {
		this.data.name = this.data.name.replace(stylingRegex, "").trim();
	}

	public hasChampionTime() {
		return this.data.championTime > 0;
	}

	public hasTrackmaniaExchangeId() {
		return this.data.tmxId !== null;
	}

	public getMedal(value: number | TrackmaniaTime) {
		const time = typeof value === "number" ? value : value.getTotalInMilliSeconds();

		const { championTime, authorTime, goldTime, silverTime, bronzeTime } = this.data;

		if (this.hasChampionTime() && time <= championTime) {
			return Medal.CHAMPION;
		}

		if (time <= authorTime) {
			return Medal.AUTHOR;
		}

		if (time <= goldTime) {
			return Medal.GOLD;
		}

		if (time <= silverTime) {
			return Medal.SILVER;
		}

		if (time <= bronzeTime) {
			return Medal.BRONZE;
		}

		return Medal.NONE;
	}

	public getDifferenceToMedal(medal: Medal, value: number | TrackmaniaTime) {
		const time = typeof value === "number" ? value : value.getTotalInMilliSeconds();
		const { championTime, authorTime, goldTime, silverTime, bronzeTime } = this.data;

		switch (medal) {
			case Medal.CHAMPION:
				return time - championTime;
			case Medal.AUTHOR:
				return time - authorTime;
			case Medal.GOLD:
				return time - goldTime;
			case Medal.SILVER:
				return time - silverTime;
			case Medal.BRONZE:
				return time - bronzeTime;
			default:
				throw new Error("Not possible to get the difference to this medal.");
		}
	}

	public getMedalTime(medal: Medal) {
		const { championTime, authorTime, goldTime, silverTime, bronzeTime } = this.data;

		switch (medal) {
			case Medal.CHAMPION:
				return new TrackmaniaTime(championTime);
			case Medal.AUTHOR:
				return new TrackmaniaTime(authorTime);
			case Medal.GOLD:
				return new TrackmaniaTime(goldTime);
			case Medal.SILVER:
				return new TrackmaniaTime(silverTime);
			case Medal.BRONZE:
				return new TrackmaniaTime(bronzeTime);
			default:
				throw new Error("Not possible to get the time of this medal.");
		}
	}

	public get tmxLink() {
		const { tmxId } = this.data;
		if (tmxId === null) return undefined;
		return `https://trackmania.exchange/maps/${tmxId}`;
	}

	public get trackmaniaIOLink() {
		const { uid } = this.data;
		if (uid === null) return undefined;
		return `https://trackmania.io/#/leaderboard/${uid}`;
	}

	public toString() {
		const hasChampionTime = this.hasChampionTime();
		const hasTrackmaniaExchangeId = this.hasTrackmaniaExchangeId();

		const { author, name } = this.data;

		const tmx = this.tmxLink;
		const champion = this.getMedalTime(Medal.CHAMPION).toString();
		const authorMedal = this.getMedalTime(Medal.AUTHOR).toString();

		if (hasTrackmaniaExchangeId && hasChampionTime) {
			assertIsDefined(tmx, "TrackmaniaExchange Link");
			return `"${name}" by ${author} | ${Emote.CHAMPION_MEDAL.name}  ${champion} | ${Emote.AUTHOR_MEDAL.name}  ${authorMedal} | ${tmx} | ${this.trackmaniaIOLink}`;
		}

		if (!hasTrackmaniaExchangeId && hasChampionTime) {
			return `"${name}" by ${author} | ${Emote.CHAMPION_MEDAL.name}  ${champion} | ${Emote.AUTHOR_MEDAL.name}  ${authorMedal} | ${this.trackmaniaIOLink}`;
		}

		if (hasTrackmaniaExchangeId && !hasChampionTime) {
			assertIsDefined(tmx, "TrackmaniaExchange Link");
			return `"${name}" by ${author} | ${Emote.AUTHOR_MEDAL.name}  ${authorMedal} | ${tmx} | ${this.trackmaniaIOLink}`;
		}

		return `"${name}" by ${author} | ${Emote.AUTHOR_MEDAL.name}  ${authorMedal} | ${this.trackmaniaIOLink}`;
	}
}

File: pkg/trackmania/map.go
package trackmania

import (
	"fmt"
	"regexp"

	"github.com/SoWieMarkus/wheelgpt/pkg/emote"
)

// stylingRegex is a regular expression that matches Trackmania styling codes in map names.
// See: https://wiki.trackmania.io/en/content-creation/text-styling
var stylingRegex = regexp.MustCompile(`\$([wnoitsgz$WNOITSGZ]|[0-9A-Fa-f]{3})`)

// Map represents a Trackmania map with its unique identifier, name, author, and medal times.
type Map struct {
	// UID is the unique identifier of a map.
	UID string

	// Name is the name of the map.
	Name string

	// Author is the name of the creator of a map.
	Author string

	// AuthorTime is the time that is required for author medal of a map.
	AuthorTime Time
	// GoldTime is the time that is required for gold medal of a map.
	GoldTime Time
	// SilverTime is the time that is required for silver medal of a map.
	SilverTime Time
	// BronzeTime is the time that is required for bronze medal of a map.
	BronzeTime Time

	// ChampionTime is the time that is required for champion medal of a map. Since this is only accessible through an OpenPlanet plugin this field is optional and may be nil if the information is not available.
	ChampionTime *Time

	// TmxID is the unique identifier of a map on Trackmania Exchange. Since not every map is available on Trackmania Exchange this field is optional and may be nil if the information is not available.
	TmxID *int64
}

// CleanName returns the name of the map with all trackmania styling codes removed.
// See: https://wiki.trackmania.io/en/content-creation/text-styling
func (m *Map) CleanName() string {
	return stylingRegex.ReplaceAllString(m.Name, "")
}

// GetMedal returns the medal that corresponds to the given time for the map.
func (m *Map) Medal(time Time) Medal {
	switch {
	case m.ChampionTime != nil && time <= *m.ChampionTime:
		return MedalChampion
	case time <= m.AuthorTime:
		return MedalAuthor
	case time <= m.GoldTime:
		return MedalGold
	case time <= m.SilverTime:
		return MedalSilver
	case time <= m.BronzeTime:
		return MedalBronze
	default:
		return MedalNone
	}
}

// TimeDifferenceToMedal returns the time difference between the given time and the required time for the specified medal on the map.
// If the medal is not available for the map, it returns 0.
func (m *Map) TimeDifferenceToMedal(medal Medal, time Time) Time {
	switch medal {
	case MedalChampion:
		if m.ChampionTime == nil {
			return 0
		}
		diff := *m.ChampionTime - time
		return diff
	case MedalAuthor:
		diff := m.AuthorTime - time
		return diff
	case MedalGold:
		diff := m.GoldTime - time
		return diff
	case MedalSilver:
		diff := m.SilverTime - time
		return diff
	case MedalBronze:
		diff := m.BronzeTime - time
		return diff
	default:
		return 0
	}
}

// String returns a string representation of the map, including its name, author, medal times, and links to Trackmania.io and Trackmania Exchange (if available).
func (m *Map) String() string {
	name := m.CleanName()

	trackmaniaIOLink := fmt.Sprintf("https://trackmania.io/#/leaderboard/%s", m.UID)
	hasChampionTime := m.ChampionTime != nil
	hasTmxID := m.TmxID != nil
	tmxLink := ""
	if hasTmxID {
		tmxLink = fmt.Sprintf("https://trackmania.exchange/map/%d", *m.TmxID)
	}

	switch {
	case hasChampionTime && hasTmxID:
		return fmt.Sprintf(`"%s" by %s | %s %s | %s %s | %s | %s`,
			name, m.Author,
			emote.ChampionMedal, m.ChampionTime,
			emote.AuthorMedal, m.AuthorTime,
			trackmaniaIOLink, tmxLink,
		)
	case hasChampionTime && !hasTmxID:
		return fmt.Sprintf(`"%s" by %s | %s %s | %s %s | %s`,
			name, m.Author,
			emote.ChampionMedal, m.ChampionTime,
			emote.AuthorMedal, m.AuthorTime,
			trackmaniaIOLink,
		)
	case !hasChampionTime && hasTmxID:
		return fmt.Sprintf(`"%s" by %s | %s %s | %s | %s`,
			name, m.Author,
			emote.AuthorMedal, m.AuthorTime,
			trackmaniaIOLink, tmxLink,
		)
	default:
		return fmt.Sprintf(`"%s" by %s | %s %s | %s`,
			name, m.Author,
			emote.AuthorMedal, m.AuthorTime,
			trackmaniaIOLink,
		)
	}
}
